### PR TITLE
CanMakePaymentEvent

### PR DESCRIPTION
**Pull request by rsolomakhin**

### CODE REVIEW

**Comment by rsolomakhin** [comment]
@marcoscaceres FYI. Going to address your comments here.

**Comment by rsolomakhin** [comment]
@marcoscaceres, please take a look. I've added permission check and removed the `canMakePaymentResult = false` line that concerned you.

**Comment by w3c-bots** [comment]
# Build PENDING

Started: None
Finished: None





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/21808)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/316088942)

**Comment by marcoscaceres** [comment]
So, everything looks good with regards to the test - but they only test a subset of what the IDL defines. We can definitely merge this in parts, but it would be good to have a plan to fully test everything that is being defined in the IDL. I worry that otherwise we will land this and things will fall through the cracks.

**Comment by rsolomakhin** [comment]
I can add more tests that check `topLevelOrigin`, `paymentRequestOrigin`, `methodData`, and `modifiers`. Is that what you had in mind?

**Comment by marcoscaceres** [comment]
@rsolomakhin, Yeah, exactly. Should we do that as a different PR? I don't mind it being part of this one tho, as I've already reviewed what is already here.

**Comment by rsolomakhin** [comment]
Let's do it here. Brb.

**Comment by rsolomakhin** [comment]
@marcoscaceres, please take a look.

**Comment by marcoscaceres** [comment]
@rsolomakhin, for the worker, I've reduced it down a bit and cleaned up the errors - we should't need "worker says:", as the Error of the promise will contain the stack trace back to the Worker. Might be nicer to also use an actual URL for URL comparisons... all included below. 

WDYT?

```JS
self.addEventListener("canmakepayment", event => {
  if (event.methodData.length !== 1) {
    const msg = "Expected exactly one method data.";
    event.respondWith(Promise.reject(new Error(msg)));
    return;
  }

  const [method] = event.methodData;
  if (!method || method.supportedMethods.length !== 1) {
    const msg = "Expected exactly one supported method name";
    event.respondWith(Promise.reject(new Error(msg)));
    return;
  }

  if (method.data.defaultParameter !== "defaultValue") {
    const msg = `Unexpected value for "defaultParameter": ${
      method.data.defaultParameter
    }`;
    event.respondWith(Promise.reject(new Error(msg)));
    return;
  }

  if ("defaultUnsupportedParameter" in method.data) {
    const msg = 'Unexpected "defaultUnsupportedParameter"';
    event.respondWith(Promise.reject(new Error(msg)));
    return;
  }

  for (const attribute of ["total", "displayItems"]) {
    if (attribute in event) {
      const msg = `Unexpected "${attribute}" attribute in CanMakePaymentEvent.`;
      event.respondWith(Promise.reject(new Error(msg)));
      return;
    }
  }

  if (event.modifiers.length !== 1) {
    const msg = "Expected exactly one modifier";
    event.respondWith(Promise.reject(new Error(msg)));
    return;
  }

  const [modifier] = event.modifiers;

  if (!modifier || modifier.supportedMethods.length !== 1) {
    const msg = "Expected exactly one supported method name in modifier";
    event.respondWith(Promise.reject(new Error(msg)));
    return;
  }

  for (const member of [
    "additionalDisplayItems",
    "modifiedUnsupportedParameter",
    "total",
  ]) {
    if (member in modifier) {
      const msg = `Unexpected member "${member}" in modifier`;
      event.respondWith(Promise.reject(new Error(msg)));
      return;
    }
  }

  const [methodName] = method.supportedMethods;
  if (methodName === "basic-card") {
    const msg =
      '"basic-card" payment method must never be checked in CanMakePaymentEvent';
    event.respondWith(Promise.reject(new Error(msg)));
    return;
  }

  const [modifierMethodName] = modifier.supportedMethods;
  if (modifierMethodName !== methodName) {
    const msg = `Unexpected modifier method name: "${modifierMethodName}". Expected "${methodName}".`;
    event.respondWith(Promise.reject(new Error(msg)));
    return;
  }

  if (modifier.data.modifiedParameter !== "modifiedValue") {
    const msg = `Unexpected value for 'modifiedParameter': ${
      modifier.data.modifiedParameter
    }`;
    event.respondWith(Promise.reject(new Error(msg)));
    return;
  }

  const methodAsURL = new URL(methodName);
  if (event.topLevelOrigin !== methodAsURL.origin) {
    const msg = `Unexpected event.topLevelOrigin: "${
      event.topLevelOrigin
    }". Expected "${methodAsURL.origin}".`;
    event.respondWith(Promise.reject(new Error(msg)));
    return;
  }

  if (event.paymentRequestOrigin !== methodAsURL.origin) {
    const msg = `Unexpected iframe origin ${event.paymentRequestOrigin}`;
    event.respondWith(Promise.reject(new Error(msg)));
    return;
  }

  switch (methodAsURL.pathname.substr(1)) {
    case "canMakePayment-true":
      event.respondWith(true);
      break;
    case "canMakePayment-false":
      event.respondWith(false);
      break;
    case "canMakePayment-promise-true":
      event.respondWith(Promise.resolve(true));
      break;
    case "canMakePayment-promise-false":
      event.respondWith(Promise.resolve(false));
      break;
    case "canMakePayment-custom-error":
      event.respondWith(Promise.reject(new Error("Custom error")));
      break;
    default:
      const msg = `Unrecognized payment method name "${methodName}".`;
      event.respondWith(Promise.reject(new Error(msg)));
      break;
  }
});
```

File: payment-handler/can-make-payment-event-constructor.https.worker.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  importScripts('/resources/testharness.js');
+}
+
+// https://w3c.github.io/payment-handler/#the-canmakepaymentevent
+
+test(() => {
+  try {
+    new CanMakePaymentEvent('test');
+  } catch (err) {
+    assert_true(false, `Unexpected exception: ${err.message}`);
```
Comment (marcoscaceres) [inline payment-handler/can-make-payment-event-constructor.https.worker.js]: nit: `assert_unreached(err.message)` instead.

File: payment-handler/can-make-payment-event-constructor.https.worker.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
+'use strict';
+
+if (self.importScripts) {
+  importScripts('/resources/testharness.js');
+}
+
+// https://w3c.github.io/payment-handler/#the-canmakepaymentevent
```
Comment (marcoscaceres) [inline payment-handler/can-make-payment-event-constructor.https.worker.js]: nit: move to top of file

File: payment-handler/can-make-payment-event-constructor.https.worker.js (hunk 1 above)
Comment (marcoscaceres) [reply payment-handler/can-make-payment-event-constructor.https.worker.js]: (yes, I should fix this in the other tests :::))

**Comment by rsolomakhin** [comment]
@marcoscaceres , please take another look. I took your comments and ran them through prettier.js, which, most noticeably, replaced double-quotes with single quotes in JS.

**Comment by w3c-bots** [comment]
# Build PENDING

Started: None
Finished: None





View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/23294)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/333387192)

File: payment-handler/can-make-payment-event-constructor.https.worker.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    composed: true,
+  });
+  assert_false(ev.isTrusted, 'constructed in script, so not be trusted');
+  assert_true(ev.bubbles, 'set by EventInitDict');
+  assert_true(ev.cancelable, 'set by EventInitDict');
+  assert_true(ev.composed, 'set by EventInitDict');
+  assert_equals(ev.target, null, 'initially null');
+  assert_equals(ev.type, 'test');
+}, 'CanMakePaymentEvent can be constructed with an EventInitDict, even if not trusted');
+
```
Comment (marcoscaceres) [inline payment-handler/can-make-payment-event-constructor.https.worker.js]: Sorry, there is one more set of tests missing here. You are testing `EventInit`, but not `CanMakePaymentEventInit`. I was expecting a test like:

```JS
test(() => {
  const ev = new CanMakePaymentEvent('test', {
    topLevelOrigin: "https://foo.com",
    paymentRequestOrigin: "https://bar.com",
    methodData: [],
    modifiers: [],
  });
  assert_false(ev.isTrusted, 'constructed in script, so not be trusted');
  assert_equals(ev.topLevelOrigin, "https://foo.com");
  assert_equals(ev.paymentRequestOrigin, "https://bar.com");
}, 'CanMakePaymentEvent can be constructed with an CanMakePaymentEventInit, even if not trusted');
```

And variations of the above, like providing weird `topLevelOrigin` values, and so on.

**Review by marcoscaceres — COMMENTED** [review]
These are looking good, @rsolomakhin. Nearly there. Just had a question.

**Comment by rsolomakhin** [comment]
Can you share your prettier config, @marcoscaceres ?

**Comment by rsolomakhin** [comment]
@marcoscaceres : Sorry about making your write the test through this code review! Please take another look.

File: payment-handler/app-can-make-payment.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    return;
+  }
+
+  if ('defaultUnsupportedParameter' in method.data) {
+    const msg = 'Unexpected "defaultUnsupportedParameter"';
+    event.respondWith(Promise.reject(new Error(msg)));
+    return;
+  }
+
+  for (const attribute of ['total', 'displayItems']) {
```
Comment (marcoscaceres) [inline payment-handler/app-can-make-payment.js]: I'm actually not sure we need this test... this should be covered by the idlHarness tests.

Comment (rsolomakhin) [reply payment-handler/app-can-make-payment.js]: Removed. Please take another look.

**Comment by rsolomakhin** [comment]
Removed the manual checks. Please take another look, @marcoscaceres.

**Review by marcoscaceres — APPROVED** [review]
Lgtm... but we should come back to do those additional event init tests I mentioned in a different PR.

**Comment by rsolomakhin** [comment]
> And variations of the above, like providing weird `topLevelOrigin` values, and so on.

Did you mean things like null characters, very long strings, and invalid URLs?

**Comment by marcoscaceres** [comment]
Yep, or is that UA contolled? If it is, then never mind :)

**Comment by rsolomakhin** [comment]
It is UA controlled, but still good idea to poke at it with a stick. brb.

**Comment by marcoscaceres** [comment]
Ok, but feel free to send as a new PR if you want to merge this one.

**Pull request merged**